File: 0946-validate-stack-sequences/0946-validate-stack-sequences.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> s;
        int i = 0;
        for (const int &val : pushed) {
            s.emplace(val);
            while (!s.empty() and s.top() == popped[i]) {
                ++i;
                s.pop();
            }
        }
        return s.empty();
    }
};

File: 0946-validate-stack-sequences/0946-validate-stack-sequences.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        var s = new Stack<Integer>();
        int i = 0;
        for (int val : pushed) {
            s.push(val);
            while (!s.empty() && s.peek() == popped[i]) {
                ++i;
                s.pop();
            }
        }
        return s.isEmpty();
    }
}